network/errors.go

```diff
@@ -2,8 +2,39 @@ package network
 
 import (
 	"errors"
+	"fmt"
 )
 
 var (
 	EmptyTargetList = errors.New("target list empty")
 )
+
+// TransientError represents an error returned from a network layer function call
+// which may be interpreted as non-critical. In general, we desire that all expected
+// error return values are enumerated in a function's documentation - any undocumented
+// errors are considered fatal. However, 3rd party libraries don't always conform to
+// this standard, including the networking libraries we use. This error type can be
+// used to wrap these 3rd party errors on the boundary into flow-go, to explicitly
+// mark them as non-critical.
+type TransientError struct {
```

Comment:
It's definitely not a good practice to do this when critical vs non-critical errors can be programmatically differentiated. But not every dependency we use returns errors that can be parsed this way. The reason for introducing this error type is to make interactions with these dependencies compatible with our error handling approach as much as possible. 

We already treat [eg. networking errors](https://github.com/onflow/flow-go/blob/ccbd21bee9f1f9fab8b1fdb72db3cb66a8229650/engine/collection/compliance/engine.go#L367-L371) like this regularly, by interpreting generic errors as non-critical, like:
```go
err := conduit.Unicast(...)
if err != nil {
    log.Err(err).Msg("failed to send message)
    return nil
}
```

If we wrap these errors with a documented type (`TransientError`), we can at least:
* handle unexpected errors further up the stack
* retry on networking failures
For example:
```go
// Error returns:
// - TransientError if dependencyWithoutDefinedErrors returns any error
// - generic error in case of unexpected 
func DoSomethingIncludingANetworkCall() error {
    err := dependencyWithoutDefinedErrors.SomethingThatMightTransientlyFail(...)
    if err != nil {
        return NewTransientError(err)
    }

    err := internalComponent.DoSomething()
    if err != nil {
        return fmt.Errorf("unexpected error in internalComponent: %w", err)
    }

    return nil
}
```

If we don't wrap the first error above, then we can't differentiate between the generic error from `dependencyWithoutDefinedErrors` (in the networking layer we generally want to treat these as non-critical) and the generic error form `internalComponent` (we must crash the component or node if we observe this).

Reply:
I agree that this is useful, and we probably want to use it, but I would like to see some checks for _some_ errors. I mean I am not very comfortable when we swallow _all_ errors without checking what has exactly happened.

Reply:
After our discussion, I added comments to each point where we wrap the error saying that we're treating it as non-critical.